predicators/datasets/generate_atom_trajs_with_vlm.py

```diff
@@ -958,16 +1008,72 @@ def create_ground_atom_data_from_generated_demos(
+                    # arguments.
+                    prev_state = curr_traj_states_for_vlm[i - 1]
+                    assert prev_state.simulator_state is not None
+                    assert "images" in prev_state.simulator_state
+                    relevant_states_for_crop = [state, prev_state]
+                    if i != 0:
+                        # Figure out which cells to include in the crop.
+                        action = curr_traj_actions_for_vlm[i - 1].get_option()
+                        min_col = env.num_cols
+                        max_col = 0
+                        min_row = env.num_rows
+                        max_row = 0
+                        # Get the (x, y) position of each relevant object.
+                        for o in action.objects:
+                            for s in relevant_states_for_crop:
+                                row = s.get(o, "row")
+                                col = s.get(o, "col")
+                                min_col = int(min(min_col, col))
+                                max_col = int(max(max_col, col))
+                                min_row = int(min(min_row, row))
+                                max_row = int(max(max_row, row))
+                        # Count rows from the bottom rather than the top,
+                        # because we actually want to index by traditional
+                        # (x, y) in the numpy array.
+                        temp = min_row
+                        min_row = env.num_rows - 1 - max_row
+                        max_row = env.num_rows - 1 - temp
+                        # Assume that we can generate the intended crop from the
+                        # first image in state.simulator_state["images"]
+                        full_curr_img = state.simulator_state["images"][0]
+                        full_prev_img = prev_state.simulator_state["images"][0]
+                        approx_cell_size = full_curr_img.shape[
+                            0] // env.num_rows
+                        cropped_curr_img = full_curr_img[
+                            min_row * approx_cell_size:(max_row + 1) *
+                            approx_cell_size,
+                            min_col * approx_cell_size:(max_col + 1) *
+                            approx_cell_size, :]
+                        cropped_prev_img = full_prev_img[
+                            min_row * approx_cell_size:(max_row + 1) *
+                            approx_cell_size,
+                            min_col * approx_cell_size:(max_col + 1) *
+                            approx_cell_size, :]
+                        cropped_imgs = [
+                            PIL.Image.fromarray(img_arr)  # type: ignore
+                            for img_arr in
+                            [cropped_curr_img, cropped_prev_img]
+                        ]
+                        cropped_state_imgs.append(cropped_imgs)
```

Comment:
important: raise a `NotImplementedError` if the `CFG.env` isn't one of these for which we've implemented cropping!